docs/src/content/v1-migration/groupBy.md

```diff
@@ -0,0 +1,90 @@
+# Typing
+
+## Strict
+
+If you weren't using the `strict` variant, the return type is now stricter by
+using the type returned from the grouping function as the type for the keys of
+the result (this is the same as the built-in [`Object.groupBy`](https://github.com/microsoft/TypeScript/blob/main/src/lib/esnext.object.d.ts#L7-L10)). To maintain the
+previous return value make sure your function returns `PropertyKey` instead of
+a narrower type.
+If you are already using `strict` simply remove the `.strict` suffix.
+
+## Indexed
+
+The `indexed` variant was removed; the base implementation takes the same
+parameters. If you are using `indexed` you can simply remove it without any
+other changes.
+
+# Runtime
+
+The grouping function now takes 2 additional parameters: `index` - The index of
+the current element being processed in array, and `data` - the array the
+function was called upon (_unlike_ the signature of the built-in
+[`Object.groupBy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/groupBy), which does't take the 3rd param).
+
+If you are using a function reference for the predicate (and not an inline arrow
+function), and that function accepts more than one param you might run into
+compile-time (or run-time!) issues because of the extra params being sent on
+each invocation of the function. We highly recommend using [unicorn/no-array-callback-reference](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-array-callback-reference.md)
+to warn against these issues.
+
+## Examples
+
+### Strict variant removed
+
+```ts
+// Was
+groupBy.strict(array, groupingFunc);
+
+// Now
+groupBy(array, groupingFunc);
+```
+
+### Indexed variant removed
+
+```ts
+// Was
+groupBy.indexed(array, groupingFunc);
+
+// Now
+groupBy(array, groupingFunc);
+```
+
+### Better typing
+
+```ts
+type Pet = { type: "cat" | "dog"; name: string };
+
+const grouped = groupBy(
+  //  ^? Partial<Record<"cat" | "dog", NonEmptyArray<Pet>>>
+  // Was: Record<PropertyKey, NonEmptyArray<Pet>>
```

Comment:
**thought:** should we change the order of these?

i was confused at first that the was comes after the now (which is not the case for the examples, but is the case for the types and results). maybe it'd be better to be consistent? but i'm not sure. anyway, we can change this all together if we decide it's clearer

Reply:
I put the new type first in all the type examples where I use the `^?` sign and add the "was" afterward. I started doing it intuitively, but now it's everywhere. It's because I thought the expectation from the `^?` would be that it would tell you the existing type, and the "was: " is like a comment on top of it.

wdyt?

Reply:
yeah i'm not sure... i agree that it's a bit late since we already have other type examples...

i was thinking we could have something like

```ts
  //  ^?
  // Was: Record<PropertyKey, NonEmptyArray<Pet>>
  // Now: Partial<Record<"cat" | "dog", NonEmptyArray<Pet>>>
```

it takes more vertical space, but i also dk how much that's a consideration since a lot of our migration notes are already long lol

Reply:
but also if it was like an inline comment it'd be like

```ts
const result = doSomething(data); // => Was: string | undefined; Now: string
```

or

```ts
const result = doSomething(data); // => Was: string | undefined
                                  //    Now: string
```

Reply:
i don't like any of these solutions, so i'm fine with how it is now—if we change our minds later, i can take point on changing everything

Reply:
Let's defer this to later if we get feedback that the docs aren't good